rfcs/0000-operator-managed-apim-config.md

```diff
@@ -0,0 +1,223 @@
+kapi ->> ctrl: Notifies about new/changed ProductAPI
+ctrl ->> kapi: Read desired ProductApi specification
+alt if actual state does not match desired state
+  ctrl ->> apim: Read actual state in Api management
+  ctrl ->> apim: Update state in Api management to match desired state
+  ctrl ->> apim: Check status of update job
+  ctrl ->> kapi: Update status of ProductApi
+end
+```
+
+Sequence diagram sync loop
+```mermaid
+sequenceDiagram
+participant dev as Developer
+participant kapi as Kubernetes API
+participant ctrl as Operator
+participant apim as Api Management
+kapi ->> ctrl: Reconcile of ProductApi
+ctrl ->> kapi: Read desired ProductApi specification
+ctrl ->> apim: Read actual state in Api management
+alt if actual state does not match desired state
+  ctrl ->> apim: Update state in Api management to match desired state
+  ctrl ->> apim: Check status of update job
+  ctrl ->> kapi: Update status of ProductApi
+end
+```
+
+Sequence diagram status api
+```mermaid
+sequenceDiagram
+participant dev as Developer
+participant api as StatusApi (custom)
+participant kapi as Kubernetes API
+dev ->> api: Get status for deployment
+api ->> kapi: Get status of resouce in kubernetes
+kapi ->> api: Returns current status of resource(s)
+api ->> dev: Returns current status of deployment
+```
+
+# Drawbacks
+
+* We need to learn and understand the kubernetes operator pattern
+* Tied to kubernetes as a "provider"/"system"
+* Faults in the operator will affect all the APIs it manages
+
+# Rationale and alternatives
+
+* Terraform modules that each team deploys
+* Bicep modules that each team deploys
+* Centralized IaC configuration
+* Scripts and pipelines
+* Azure service operator
+
+# Prior art
+
+We have a pipeline script setup today that configures the api management service. This is complex and has alot of configuration elements that most of the developers shouldn't have to care about.
+
+Few people know how this works and it is complicated to deploy a new or a change to a api in the api management service
+
+# Unresolved questions
+
+* Can we automate even more of the process than we do today
+* How we should report status of an deployment back to the teams
+  * Flux status?
+  * Custom api endpoint for viewing status (ala. kuberneteswrapper)
+  * other
+
+# Future possibilities
+
+Having a operator pattern in front of Api Management can in the future automate the subscription registration for apps at run time. This will simplify our deployment process of apps.
+
+If the application is deployed in one of our kubernetes clusters we can tie the api setup tighter together with the deployment of the application.
+
+Managing subscriptions for external entities, this is a bigger decission as we need a place where customers can manage their subscriptions. We need to decide if we are going to use a SaaS service for this or write our own. Either way this operator might be a good place to handle backup/restore functionality
```

Comment:
Unsure what this means. Are you thinking of a sort of **service catalog**?

Reply:
Ah yes, that. We've thought about options here, including backstage and the aforementioned from azure. But, yes, sounds like a different product.